server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,362 @@
+        auditEventList =
+            createAuditEventCore(
+                resources, BalpProfileEnum.PATIENT_UPDATE, BalpProfileEnum.PATIENT_UPDATE);
+        break;
+
+      case DELETE:
+
+        // NOTE: The success of processing of DELETE is heavily dependent on server validation
+        // policy e.g. If you have a permission list that references the resource being deleted,
+        // it breaks the delete operation. Likewise if you reference the resource in AuditEvent
+        // such as logging when you first created or updated it the operation breaks. In case of
+        // such an error, AuditEvent.outcome and AuditEvent.outcomeDesc are populated accordingly
+
+        // Also note, this implementation only fully captures that a specific resource was deleted
+        // and by whom but not the compartment owner.
+        // With no access to the actual deleted resource we can't get the compartment owner unless
+        // the resource itself is a Patient. A crude way to get the owner would be to fetch the
+        // actual resource from the database first before creating the AuditEvent.
+
+        // This implementation skips logging Deletes if the operation is conditional (thus no
+        // Resource ID present)
+
+        auditEventList =
+            createAuditEventCore(
+                resources, BalpProfileEnum.PATIENT_DELETE, BalpProfileEnum.BASIC_DELETE);
+        break;
+
+      default:
+        break;
+    }
+
+    // TODO Investigate bulk saving (batch processing) instead to improve performance. We'll
+    // probably need a mechanism for chunking e.g. 100, 200, or 500 batch
+    for (AuditEvent auditEvent : auditEventList) {
+      auditEvent.getPeriod().setEnd(new Date());
+      try {
+        this.httpFhirClient.postResource(auditEvent);
+      } catch (IOException exception) {
+        ExceptionUtil.throwRuntimeExceptionAndLog(logger, exception.getMessage(), exception);
+      }
+    }
+  }
+
+  private String getResourceTemplate(String resourceName, String resourceId) {
+    return String.format("{ \"resourceType\": \"%s\", \"id\": \"%s\"}", resourceName, resourceId);
+  }
+
+  private List<AuditEvent> createAuditEventCore(
+      List<IBaseResource> resources, BalpProfileEnum patientProfile, BalpProfileEnum basicProfile) {
+    List<AuditEvent> auditEventList = new ArrayList<>();
+
+    for (IBaseResource iBaseResource : resources) {
+      try {
+        Preconditions.checkState(iBaseResource instanceof DomainResource);
+
+        DomainResource resource = (DomainResource) iBaseResource;
+
+        Set<String> patientIds = patientFinder.findPatientIds(resource);
+
+        if (!patientIds.isEmpty()) {
+          auditEventList.add(createAuditEvent(resource, patientProfile, patientIds));
+        } else {
+          auditEventList.add(createAuditEvent(resource, basicProfile, Set.of()));
+        }
+      } catch (IllegalStateException exception) {
+        logger.error(exception.getMessage(), exception);
```

Comment:
Who throws this exception and why are we not rethrowing it?

Reply:
This is thrown by the line here checking the type of resource we are processing - https://github.com/google/fhir-gateway/blob/d4bd38fcf7002554534da545e60edb56993243d1/server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java#L155

We are not rethrowing so that we don't terminate processing for the valid states.

Reply:
Refactored this to an if/else condition and removed the try catch.